### PR TITLE
Add the Gluon Implementation of Deformable Convolution

### PR DESCRIPTION
**Pull request by suyz526**
## Description ##
The gluon implementation of deformable convolution

### Essentials ###
-  Changes are complete (i.e. I finished coding on this PR)
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- For user-facing API changes, API doc string has been updated.

### CODE REVIEW

**Comment by suyz526** [comment]
@ChaiBapchya @sandeep-krishnamurthy @vandanavk @aaronmarkham @larroy

Hi, 
things became complex after rebase...So I moved here with a new request.
I really have no idea about the Website error.

**Comment by suyz526** [comment]
http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

and where can I find the `BUILD_ID` for a preview?

File: python/mxnet/gluon/contrib/cnn/conv_layers.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    Outputs:
+        - **out**: 4D output tensor with shape
+          `(batch_size, channels, out_height, out_width)` when `layout` is `NCHW`.
+          out_height and out_width are calculated as::
+
+              out_height = floor((height+2*padding[0]-dilation[0]*(kernel_size[0]-1)-1)/stride[0])+1
+              out_width = floor((width+2*padding[1]-dilation[1]*(kernel_size[1]-1)-1)/stride[1])+1
+
+    Reference:
+        .. [1] Dai, Jifeng, et al. "Deformable convolutional networks." CoRR, abs/1703.06211 1.2 (2017): 3.
```
Comment (aaronmarkham) [inline python/mxnet/gluon/contrib/cnn/conv_layers.py]: My guess about the Sphinx error is that it doesn't like this line.
Referring to the [docs on citations](https://www.sphinx-doc.org/en/1.5/rest.html#citations), you can call it what you like, so maybe pick something more descriptive than "1" and maybe you won't have a collision. It actually says to use a non-numeric label.
Maybe to align with how this is categorized in the rst docs, call this section Citations, and then when anyone looks it up we know how it is supposed to work...

Comment (suyz526) [reply python/mxnet/gluon/contrib/cnn/conv_layers.py]: Now I simplified the citations as it is in Gluon BN layer: 
https://mxnet.incubator.apache.org/api/python/gluon/nn.html#mxnet.gluon.nn.BatchNorm

Besides, in the doc of Gluon.Contrib syncbatchnorm:
https://mxnet.incubator.apache.org/api/python/gluon/contrib.html?highlight=syncbatchnorm#id3

it seems to be able to create a hyperlink with an underline, e.g. [1]_ 

Now the website error is gone, thanks very much for your help.

**Comment by suyz526** [comment]
@mxnet-label-bot update [Gluon, pr-awaiting-review]

**Comment by aaronmarkham** [comment]
> http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
> 
> and where can I find the `BUILD_ID` for a preview?

If you click on "Show all checks" in the CI zone of this PR, you'll find one called "ci/jenkins/mxnet-validation/website". Click details. In the deploy stage (look for the green check marks), you'll see Julia selected. Select "Docs". Then click the last shell script. Then look all the way at the bottom of the output. There's your new link for a preview.

**Comment by aaronmarkham** [comment]
Latest preview: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-14810/5/index.html

**Comment by suyz526** [comment]
> > http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
> > and where can I find the `BUILD_ID` for a preview?
> 
> If you click on "Show all checks" in the CI zone of this PR, you'll find one called "ci/jenkins/mxnet-validation/website". Click details. In the deploy stage (look for the green check marks), you'll see Julia selected. Select "Docs". Then click the last shell script. Then look all the way at the bottom of the output. There's your new link for a preview.

I see, thanks very much!

**Comment by wkcn** [comment]
Merged. Thank you!

**Comment by lucinyaLi** [comment]
I met some error when I used the deformableconv layer presented above.
The error is :
MXNetError: [14:16:07] src/operator/contrib/././nn/deformable_im2col.h:94: only implemented in GPU
And I found some code you wrote 
`ctx = mx.gpu()
net.initialize(force_reinit=True, ctx=ctx)
`
And in my code,  only 
`    if args.resume.strip():
        net.load_parameters(args.resume.strip())
        async_net.load_parameters(args.resume.strip())
    else:
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            net.initialize()
            async_net.initialize()
`
I change my code to   net.initialize(force_reinit=True, ctx=ctx), but there is another error  :
RuntimeError: Parameter 'ssd0_vggatrousextractor0_init_scale' was not initialized on context cpu(0). It was only initialized on [gpu(1)].

Can I get your help ?

**Comment by suyz526** [comment]
Hi,

currently, the deformable convolution layer only supports the computation on GPU, so you need to make sure that all the parameter are stored on GPU. This can be done with 
```
ctx = mx.gpu()  #ctx = mx.gpu(gpu_id)  
net.initialize(ctx=ctx) 
```
or in your case,
`net.load_parameters(path, ctx=ctx)`

Do you have some CustomOp in your network, which only supports CPU, e.g. it was implemented with pure numpy?

**Comment by lucinyaLi** [comment]
> CustomOp
This is a part of my net, I change some original conv layer in SSD into the deformable conv. There is nothing else I have changed. 
    with self.name_scope():
            # we use pre-trained weights from caffe, initial scale must change
            init_scale = mx.nd.array([0.229, 0.224, 0.225]).reshape((1, 3, 1, 1)) * 255
            self.init_scale = self.params.get_constant('init_scale', init_scale)
            self.stages = nn.HybridSequential()
            for l, f in zip(layers, filters):
                stage = nn.HybridSequential(prefix='')
                with stage.name_scope():
                        for _ in range(l):
                            stage.add(DeformableConvolution(f, kernel_size=3, padding=1, **self.init))
                            if batch_norm:
                                stage.add(nn.BatchNorm())
                            stage.add(nn.Activation('relu'))
                self.stages.add(stage)
and then
    net.initialize(ctx=ctx) # where ctx=gpu(1)
    net.load_parameters(path, ctx=ctx)
I think all the parameter are stored on GPU. Is it right ? But the error  :
RuntimeError: Parameter 'ssd0_vggatrousextractor0_init_scale' was not initialized on context cpu(0). It was only initialized on [gpu(1)].
What the error mean is that I should initialize the self.init_scale in CPU?

**Comment by suyz526** [comment]
> > CustomOp
> > This is a part of my net, I change some original conv layer in SSD into the deformable conv. There is nothing else I have changed.
> > with self.name_scope():
> > # we use pre-trained weights from caffe, initial scale must change
> > init_scale = mx.nd.array([0.229, 0.224, 0.225]).reshape((1, 3, 1, 1)) * 255
> > self.init_scale = self.params.get_constant('init_scale', init_scale)
> > self.stages = nn.HybridSequential()
> > for l, f in zip(layers, filters):
> > stage = nn.HybridSequential(prefix='')
> > with stage.name_scope():
> > for _ in range(l):
> > stage.add(DeformableConvolution(f, kernel_size=3, padding=1, **self.init))
> > if batch_norm:
> > stage.add(nn.BatchNorm())
> > stage.add(nn.Activation('relu'))
> > self.stages.add(stage)
> > and then
> > net.initialize(ctx=ctx) # where ctx=gpu(1)
> > net.load_parameters(path, ctx=ctx)
> > I think all the parameter are stored on GPU. Is it right ? But the error  :
> > RuntimeError: Parameter 'ssd0_vggatrousextractor0_init_scale' was not initialized on context cpu(0). It was only initialized on [gpu(1)].
> > What the error mean is that I should initialize the self.init_scale in CPU?

Try 
`
init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=ctx).reshape((1, 3, 1, 1)) * 255
`

**Comment by lucinyaLi** [comment]
@suyz526
init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.gpu(1)).reshape((1, 3, 1, 1)) * 255
Or
init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.cpu(0)).reshape((1, 3, 1, 1)) * 255
I tried the two ways.
It does no work. the error still exists.

**Comment by suyz526** [comment]
> @suyz526
> init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.gpu(1)).reshape((1, 3, 1, 1)) * 255
> Or
> init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.cpu(0)).reshape((1, 3, 1, 1)) * 255
> I tried the two ways.
> It does no work. the error still exists.

sry, this step is unnecessary. How about your input data?

**Comment by lucinyaLi** [comment]
> > @suyz526
> > init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.gpu(1)).reshape((1, 3, 1, 1)) * 255
> > Or
> > init_scale = mx.nd.array([0.229, 0.224, 0.225], ctx=mx.cpu(0)).reshape((1, 3, 1, 1)) * 255
> > I tried the two ways.
> > It does no work. the error still exists.
> 
> sry, this step is unnecessary. How about your input data?

Nonono. I am very grateful  for your help.  
Just now, I found that this error is irrelevant to the deformable conv layers. It  exists also in original SSD .
When I want to change the original code 'net = get_model(net_name, pretrained_base=True)' into 'net = get_model(net_name, pretrained_base=True, ctx=ctx)', the error exists.
And I ask this question in the SSD gluon forum, the URL is :
https://discuss.gluon.ai/t/topic/12903
I am looking forward for your further help if you want. And thanks again.

**Pull request merged**